hackathon_site/hardware/test_api.py

```diff
@@ -194,3 +194,244 @@ def test_hardware_get_success(self):
+                    "manufacturer": "manufacturer",
+                    "datasheet": "/datasheet/location/",
+                    "quantity_available": 0,
+                    "notes": "notes",
+                    "max_per_team": 1,
+                    "picture": "http://testserver/media/picture/location",
+                    "categories": [],
+                    "quantity_remaining": 0,
+                }
+            ],
+        }
+        response_false = self.client.get(url_false)
+        self.assertEqual(response_false.status_code, status.HTTP_200_OK)
+        data_false = response_false.json()
+        print(data_false)
+        self.assertEqual(data_false, expected_data_false)
+
+        # when in_stock is not present
+        url_not_present = self._build_filter_url(in_stock="")
+        expected_data_not_present = {
+            "count": 3,
+            "next": None,
+            "previous": None,
+            "results": [
+                {
+                    "id": 1,
+                    "name": "hardware1",
+                    "model_number": "model",
+                    "manufacturer": "manufacturer",
+                    "datasheet": "/datasheet/location/",
+                    "quantity_available": 0,
+                    "notes": "notes",
+                    "max_per_team": 1,
+                    "picture": "http://testserver/media/picture/location",
+                    "categories": [],
+                    "quantity_remaining": 0,
+                },
+                {
+                    "id": 2,
+                    "name": "arduino",
+                    "model_number": "model",
+                    "manufacturer": "manufacturer",
+                    "datasheet": "/datasheet/location/",
+                    "quantity_available": 4,
+                    "notes": "notes",
+                    "max_per_team": 1,
+                    "picture": "http://testserver/media/picture/location",
+                    "categories": [],
+                    "quantity_remaining": 4,
+                },
+                {
+                    "id": 3,
+                    "name": "hardware3",
+                    "model_number": "model",
+                    "manufacturer": "manufacturer",
+                    "datasheet": "/datasheet/location/",
+                    "quantity_available": 4,
+                    "notes": "notes",
+                    "max_per_team": 1,
+                    "picture": "http://testserver/media/picture/location",
+                    "categories": [],
+                    "quantity_remaining": 4,
+                },
+            ],
+        }
+        response_not_present = self.client.get(url_not_present)
+        self.assertEqual(response_not_present.status_code, status.HTTP_200_OK)
+        data_not_present = response_not_present.json()
+        print(data_not_present)
+        self.assertEqual(data_not_present, expected_data_not_present)
+
+    def test_id_filter(self):
```

Comment:
Now that we have protections against ids not being integers, you should add a test for that as well to make sure it doesn't get broken in the future!

Reply:
Still need a test for invalid ids, something like `test_invalid_ids` that does `url = self._build_filter_url(id="1,2,foo"). You can just check that `response.status_code`.

You can add this in:

```python
def test_id_invalid(self):
    self._login()

    url = self._build_filter_url(id="1,2,foo")
    response = self.client.get(url)
    data = response.json()

    self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
    self.assertEqual(data, {"id": ["Enter a whole number."]})
```